src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs

```diff
@@ -159,8 +170,19 @@ private ProjectSnapshotManagerBase GetProjectManager()
 
         protected async Task UpdateAsync(Action action)
         {
-            await CommonServices.ThreadingService.SwitchToUIThread();
-            action();
+            // The passed-in action is intended to execute solely on our specialized single-threaded
+            // foreground scheduler. However, there are times when the action may need to temporarily
+            // switch to the UI thread to perform a certain task. We want to make sure that no other
+            // action else picks up the specialized foreground thread in the meantime, so we need to
+            // protect the below area with a lock.
+            using (JoinableCollection.Join())
+            {
+                using (await _updateLock.EnterAsync().ConfigureAwait(false))
+                {
+                    await _foregroundDispatcher.SpecializedForegroundScheduler;
```

Comment:
The primary change in the PR is here (going from using the UI thread to this specialized thread)